### PR TITLE
Resolve memory issues with callStorm, et al

### PR DESCRIPTION
**Pull request by jnwatson**
The root issue is that, upon leaving a Base context manager via exception, the Exception object itself is stored on the Base. This leads to a long circular reference loop that the GC has trouble untangling.

Also add a --keep option to the benchmark_cortex.py script that keeps the benchmark test data in the cortex between runs.

Also fix a failure to validate layers when creating a view.

This speeds up coreapi.callStorm quite a bit.

### CODE REVIEW

File: synapse/lib/base.py [hunk 1]
```diff
@@ -210,20 +205,8 @@ async def __aexit__(self, exc, cls, tb):
         except RuntimeError:
             pass
 
-        self.exitok = cls is None
-        self.exitinfo = (exc, cls, tb)
```
Comment (invisig0th) [inline synapse/lib/base.py:214]: ungh, yea.  good catch!

File: synapse/lib/view.py [hunk 2]
```diff
@@ -23,9 +23,13 @@
         'name': {'type': 'string'},
         'parent': {'type': ['string', 'null'], 'pattern': s_config.re_iden},
         'creator': {'type': 'string', 'pattern': s_config.re_iden},
+        'layers': {
+            'type': 'array',
+            'items': {'type': 'string', 'pattern': s_config.re_iden}
```
Comment (invisig0th) [inline synapse/lib/view.py:28]: good catch. maybe ideally would be a different PR, but i'm guilty of this constantly so.. :D

File: scripts/benchmark_cortex.py [hunk 3]
```diff
@@ -130,36 +152,50 @@ def oe(num):
...
+            $view = $lib.view.add(($layr,))
+            $view.set(name, $name)
+            return ($view.iden)''', opts={'vars': {'name': name, 'layr': self.layriden}})
+
+        self.viewiden = retn
 
-        orgs = [(('ou:org', fredguid), {})]
-        gen = itertools.chain(self.ips, self.dnsas, self.urls, self.asns, orgs)
-        await prox.addFeedData('syn.nodes', list(gen), viewiden=self.viewiden)
+        if not already_got_one:
```
Comment (invisig0th) [inline scripts/benchmark_cortex.py:176]: alreadygotone

File: scripts/benchmark_cortex.py [hunk 4]
```diff
@@ -103,21 +123,23 @@ def oe(num):
 
         # Ip addresses with an all tag with half having an 'even' tag and the other an 'odd' tag
         self.ips = [(('inet:ipv4', ip), {'tags': {'all': (None, None), oe(ip): (None, None)}}) for ip in ips]
-        self.dnsas = [(('inet:dns:a', dnsas), {}) for dnsas in dnsas]
+        self.dnsas: List[Tuple[Tuple, Dict]] = [(('inet:dns:a', dnsas), {}) for dnsas in dnsas]
 
-        self.asns = [(('inet:asn', asn * 2), {}) for asn in range(num_records)]
-        self.asns2 = [(('inet:asn', asn * 2 + 1), {}) for asn in range(num_records)]
+        self.asns: List[Tuple[Tuple, Dict]] = [(('inet:asn', asn * 2), {}) for asn in range(work_factor)]
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply scripts/benchmark_cortex.py:128]: I only do this when mypy complains. It complains because it is unable to infer the type of the embedded dict from the assignment.

File: scripts/benchmark_cortex.py [hunk 5]
```diff
@@ -103,21 +123,23 @@ def oe(num):
...
         random.shuffle(self.asns)
         random.shuffle(self.asns2)
 
         self.asns2prop = [(asn[0], {'props': {'name': 'x'}}) for asn in self.asns]
 
-        fredguid = s_common.guid('fred')
+        fredguid = self.myguid()
         self.asns2formexist = [(asn[0], {'props': {'owner': fredguid}}) for asn in self.asns]
-        self.asns2formnoexist = [(asn[0], {'props': {'owner': '*'}}) for asn in self.asns]
+        self.asns2formnoexist = [(asn[0], {'props': {'owner': self.myguid()}}) for asn in self.asns]
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply scripts/benchmark_cortex.py:137]: Mypy *is* able to infer the type of the variable from the assignment because there's an item in the dict.
